desc/singularities.py

```diff
@@ -0,0 +1,758 @@
+            # this calculates the effect at a single evaluation point, from all others
+            # in a single field period. loop this to get all pts
+            f_temp = eval_pt_vmap(i)
+            return put(fj, i, f_temp.reshape(fj[i].shape))
+
+        # vmap for inner part found more efficient than fori_loop, especially on gpu,
+        # but for jacobian looped seems to be better and less memory
+        if loop:
+            fj = fori_loop(0, eval_grid.num_nodes, eval_pt_loop, jnp.zeros_like(f))
+        else:
+            fj = vmap(eval_pt_vmap)(jnp.arange(eval_grid.num_nodes))
+
+        f += fj.reshape((eval_grid.num_nodes, kernel.ndim))
+        return f, src_data
+
+    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
+    f, _ = fori_loop(0, src_grid.NFP, nfp_loop, (f, src_data))
+
+    # undo rotation of src_zeta
+    src_data["zeta"] = src_zeta
+    src_data["phi"] = src_phi
+    # we sum distance vectors, so they need to be in xyz for that to work
+    # but then need to convert vectors back to rpz
+    if kernel.ndim == 3:
+        f = xyz2rpz_vec(f, phi=eval_data["phi"])
+    return f
+
+
+def _singular_part(
+    eval_data,
+    eval_grid,
+    src_data,
+    src_grid,
+    s,
+    q,
+    kernel,
+    interpolator,
+    loop=False,
+):
+    """Integrate singular point by interpolating to polar grid.
+
+    Generally follows sec 3.2.2 of [2], with the following differences:
+
+    - hyperparameter M replaced by s
+    - density sigma / function f is absorbed into kernel.
+    """
+    src_dtheta = src_grid.spacing[:, 1]
+    src_dzeta = src_grid.spacing[:, 2] / src_grid.NFP
+    eval_theta = jnp.asarray(eval_grid.nodes[:, 1])
+    eval_zeta = jnp.asarray(eval_grid.nodes[:, 2])
+
+    r, w, dr, dw = _get_quadrature_nodes(q)
+
+    h_t = jnp.mean(src_dtheta)
+    h_z = jnp.mean(src_dzeta)
+
+    eta = _chi(r)
+    # integrand of eq 38 in [2] except stuff that needs to be interpolated
+    v = eta * s**2 * h_t * h_z / 4 * abs(r) * dr * dw
+    keys = list(set(["|e_theta x e_zeta|"] + kernel.keys))
+    if "phi" in keys:
+        keys += ["omega"]
+    fsrc = [src_data[key] for key in keys]
+
+    def polar_pt_vmap(i):
+        # evaluate the effect from a single polar node around each eval point
+        # on that eval point. Polar grids from other singularities have no effect
+        # See sec 3.2.2 of [2]
+        dt = s / 2 * h_t * r[i] * jnp.sin(w[i])
+        dz = s / 2 * h_z * r[i] * jnp.cos(w[i])
+        theta_i = eval_theta + dt
+        zeta_i = eval_zeta + dz
+
+        # data interpolated to each eval pt offset by dt,dz
+        src_data_polar = {key: interpolator(val, i) for key, val in zip(keys, fsrc)}
+
+        # can't interpolate phi directly since its not periodic, so we interpolate
+        # omega and add it back in
+        src_data_polar["zeta"] = zeta_i
+        src_data_polar["theta"] = theta_i
+        if "phi" in keys:
+            src_data_polar["phi"] = zeta_i + src_data_polar["omega"]
```

Comment:
for now though this effectively does nothing since we dont use omega?

Reply:
Correct, but I wanted to include the logic here for future proofing since naively interpolating phi the same way we do R,Z wouldn't work since phi itself has a secular term